cumulusci/tasks/bulkdata/utils.py

```diff
@@ -13,15 +12,66 @@
 from cumulusci.utils import convert_to_snake_case
 
 
-def batch_iterator(iterator, n=10000):
-    while True:
-        batch = list(itertools.islice(iterator, n))
-        if not batch:
-            return
+def batch_iterator(iterator, fields, n=10000, char_limit=10000000):
+    """Given an iterator of records, return batches that (in order of precedence):
+    (1) Do not exceed the given character limit
+    (2) Contain no more than n records.
+    """
 
+    per_record_csv_chars = csv_const_chars_per_record(len(fields))
+    csv_field_chars = sum_chars(fields) + per_record_csv_chars
+
+    current_chars = 0
+    current_chars += csv_field_chars
+
+    batch = []
+    for record in iterator:
+        # Does the next record put us over the character limit?
+        if sum_chars(record) + per_record_csv_chars + current_chars > char_limit:
+            yield batch
+            batch = []
+            current_chars = csv_field_chars
+
+        batch.append(record)
+        current_chars += sum_chars(record) + per_record_csv_chars
+
+        # yield batch if we're at desired size
+        if len(batch) == n:
+            yield batch
+            batch = []
+            current_chars = csv_field_chars
+
+    # give back anything leftover
+    if batch:
         yield batch
 
 
+def csv_const_chars_per_record(num_fields):
+    """Given number of fields (or values) on a record, return the
+    number of **per record** constant characters that would be applied
+    in a .csv file format.
+
+    Given fields:
+        ['Id', 'FirstName', 'LastName']
+
+    The corresponding line in a csv file would be:
+        'Id','FirstName','LastName'\n
+
+    For each field there are:
+        quotations = num_fields * 2
+        commas = num_fields - 1
+        newlines = 1
+    Some algebra gives us: 3 * num_fields
+    #TODO: Do we include spaces in between values?
+    """
+    return 3 * num_fields
+
+
+def sum_chars(list):
+    """Given a list of strings return the sum of string lengths"""
+    return len("".join(list))
```

Comment:
Wondering, given that this method may be called tens of millions of times in performance data loads, if there's a significant difference using `join()`, which presumably is doing a new string alloc, or summing the `len()` of each element.

Reply:
eliminated the need for this method :)